Dynatrace/dynatrace-operator
# Adds support for EEC PersistentVolumeClaim

[K8S-10996](https://dt-rnd.atlassian.net/browse/K8S-10996)

## Description

Transfers `dynakube.templates.extension.persistentVolumeClaim` spec to EEC statefulset. Statefulset creates `agent-runtime-dynatrace-extensions-controller-0` PersistentVolumeClaim which has to match a PersistentVolume created by a user (or automatically in GKE cluster). 

`statefulset.spec.persistentVolumeClaimRetentionPolicy` can be changed using  `dynakube.templates.extensionExecutionController.persistentVolumeClaimRetentionPolicy` field.

## How can this be tested?

Create extensions-controller-tls secret :
```
openssl req -nodes -x509 -newkey rsa:4096 -keyout key.pem -out cert.pem -days 365 -subj "/CN=dynakube-extensions-controller.dynatrace"
kubectl -n dynatrace create secret tls extensions-controller-tls --cert=cert.pem --key=key.pem
```
Create AG tlsSecretName secret (use the same export password when asked for) :
```
openssl req -nodes -x509 -newkey rsa:4096 -keyout key.pem -out cert.pem -days 365 -subj "/CN=dynakube-activegate.dynatrace"
openssl pkcs12 -export -inkey key.pem -in cert.pem -out cert.p12
kubectl -n dynatrace create secret generic dynakube-ag-tls --from-file=server.p12=cert.p12 --from-file=server.crt=cert.pem --from-literal=password=<an export password>
```
(*) Always make sure EEC app is running after dynakube is applied/changed :

$ kubectl -n dynatrace logs statefulset.apps/dynatrace-extensions-controller
```
...
2024-08-23 15:59:11.996 UTC [0000000f] info    [native] Logs ingest limits configuration arrived:
 	-MaxContentLength: 65536
 	-MaxAttributeLength: 250
... 
```	
Apply dynakube :
```
spec:
  apiUrl: https://.../api  
  activeGate:
    tlsSecretName: dynakube-ag-tls
  customPullSecret: azurecr
  extensions:
    prometheus:
      enabled: true
  templates:
    extensionExecutionController:
      imageRef:
        repository: extk8sregistry.azurecr.io/eec/dynatrace-eec
        tag: 1.301.0.20240826-070516
      tlsRefName: extensions-controller-tls
      persistentVolumeClaim:                                                                                                                                                                                         
        accessModes:                                                                                                                                                                                                 
        - ReadWriteOncePod                                                                                                                                                                                           
        storageClassName: "standard"                                                                                                                                                                                 
        resources:                                                                                                                                                                                                   
          requests:                                                                                                                                                                                                  
            storage: 1G                                                                                                                                                                                              
      persistentVolumeClaimRetentionPolicy:                                                                                                                                                                          
        whenDeleted: Delete                                                                                                                                                                                          
        whenScaled: Delete
```
Expected EEC statefulset :
```
    Mounts:
      /var/lib/dynatrace/remotepluginmodule/agent/runtime from agent-runtime (rw)
Volume Claims:
  Name:          agent-runtime
  StorageClass:  standard
  Labels:        <none>
  Annotations:   <none>
  Capacity:      1G
  Access Modes:  [ReadWriteOncePod]
```
PVC :
```
NAME                                                                    STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   VOLUMEATTRIBUTESCLASS   AGE
persistentvolumeclaim/agent-runtime-dynatrace-extensions-controller-0   Bound    pvc-1af9f031-dcb2-460d-9272-343ece9210cf   1Gi        RWOP           standard       <unset>                 9m15s
```
```
Name:          agent-runtime-dynatrace-extensions-controller-0
Namespace:     dynatrace
StorageClass:  standard
Status:        Bound
Volume:        pvc-94518e3a-1918-4024-a2cb-2ce52a76b377
Labels:        app.kubernetes.io/created-by=dynakube
               app.kubernetes.io/managed-by=dynatrace-operator
               app.kubernetes.io/name=dynatrace-extensions-controller
Annotations:   pv.kubernetes.io/bind-completed: yes
               pv.kubernetes.io/bound-by-controller: yes
               volume.beta.kubernetes.io/storage-provisioner: pd.csi.storage.gke.io
               volume.kubernetes.io/storage-provisioner: pd.csi.storage.gke.io
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      1Gi
Access Modes:  RWOP
VolumeMode:    Filesystem
Used By:       dynatrace-extensions-controller-0
```
PV created by GKE :
```
NAME                                                        CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM                                                       STORAGECLASS   VOLUMEATTRIBUTESCLASS   REASON   AGE
persistentvolume/pvc-1af9f031-dcb2-460d-9272-343ece9210cf   1Gi        RWOP           Delete           Bound    dynatrace/agent-runtime-dynatrace-extensions-controller-0   standard       <unset>                          9m11s
```
